core/App/ffi_operation.cpp

```diff
@@ -390,6 +390,283 @@ extern "C"
+                }
+     *
+     * @return char*
+     * [string] json string
+        {
+            code: int,
+            message: string,
+            result: {
+                cmk : a base64 string
+            }
+        }
+     */
+    uint32_t ffi_importKeyMaterial(JsonObj payloadJson, char *respJson)
+    {
+        RetJsonObj retJsonObj;
+
+        JsonObj tokenJsonObj;
+
+        ehsm_keyblob_t *cmk = NULL;
+        ehsm_keyblob_t *sessionkey = NULL;
+        ehsm_status_t ret = EH_OK;
+        ehsm_data_t *key_material = NULL;
+        ehsm_data_t *hmac = NULL;
+        ehsm_data_t *importToken = NULL;
+        std::string hmac_str;
+
+        JSON2STRUCT(payloadJson, cmk);
+        JSON2STRUCT(payloadJson, sessionkey);
+
+        std::string token_str = base64_decode(payloadJson.readData_string("importToken"));
+        tokenJsonObj.parse(token_str);
+
+        std::string signature = tokenJsonObj.readData_string("hmac");
+
+        Json::Value token_json = tokenJsonObj.getJson();
+        token_json.removeMember("hmac");
+        tokenJsonObj.setJson(token_json);
+        uint32_t importToken_size = tokenJsonObj.toString().size();
+        importToken = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(importToken_size));
+        importToken->datalen = importToken_size;
+
+        memcpy_s(importToken->data, importToken->datalen, tokenJsonObj.toString().c_str(), importToken->datalen);
+        importToken->data[importToken_size - 1] = 0;
+
+        string key_material_str_base64 = payloadJson.readData_string("key_material");
+
+        string key_material_str = base64_decode(key_material_str_base64);
+
+        key_material = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(key_material_str.length()));
+        key_material->datalen = key_material_str.length();
+        memcpy_s(key_material->data, key_material->datalen, key_material_str.c_str(), key_material->datalen);
+
+        ehsm_padding_mode_t padding_mode = (ehsm_padding_mode_t)payloadJson.readData_uint32("padding_mode");
+
+        // generate hmac space
+        hmac = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(EH_HMAC_SHA256_SIZE));
+        if (hmac == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+        hmac->datalen = EH_HMAC_SHA256_SIZE;
+
+        if (payloadJson.readData_string("sessionkey") == "")
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("SessionKeyBlob is empty.");
+            goto out;
+        }
+
+        ret = GenerateTokenHmac(sessionkey, importToken, hmac);
+        if (ret != EH_OK)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Generate tokenHmac failed.");
+            goto out;
+        }
+
+        hmac_str = base64_encode(hmac->data, hmac->datalen);
+        if (strcmp(hmac_str.c_str(), signature.c_str()))
```

Comment:
do not use strcmp

use something like this:
int consttime_memcmp(const void *in_a, const void *in_b, size_t len)
{
  size_t i;
  const unsigned char *a = in_a;
  const unsigned char *b = in_b;
  unsigned char x = 0;
    for (i = 0; i < len; i++)
  	x |= a[i] ^ b[i];
    return x;
  }


Or use CRYPTO_memcmp in openssl.

Reply:
this part is fixed in web.